x-pack/plugins/triggers_actions_ui/public/application/sections/rules_list/components/rule_status_dropdown.tsx

```diff
@@ -0,0 +1,471 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React, { useState, useEffect, useCallback } from 'react';
+import moment from 'moment';
+import { i18n } from '@kbn/i18n';
+import {
+  useGeneratedHtmlId,
+  EuiLoadingSpinner,
+  EuiPopover,
+  EuiContextMenu,
+  EuiBadge,
+  EuiPanel,
+  EuiFieldNumber,
+  EuiSelect,
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiButton,
+  EuiHorizontalRule,
+  EuiTitle,
+  EuiFlexGrid,
+  EuiSpacer,
+  EuiLink,
+  EuiText,
+  EuiToolTip,
+} from '@elastic/eui';
+import { parseInterval } from '../../../../../common';
+
+import { RuleTableItem } from '../../../../types';
+
+type SnoozeUnit = 'm' | 'h' | 'd' | 'w' | 'M';
+const SNOOZE_END_TIME_FORMAT = 'LL @ LT';
+
+export interface ComponentOpts {
+  item: RuleTableItem;
+  onRuleChanged: () => void;
+  enableRule: () => Promise<void>;
+  disableRule: () => Promise<void>;
+  snoozeRule: (snoozeEndTime: string | -1) => Promise<void>;
+  unsnoozeRule: () => Promise<void>;
+}
+
+export const RuleStatusDropdown: React.FunctionComponent<ComponentOpts> = ({
+  item,
+  onRuleChanged,
+  disableRule,
+  enableRule,
+  snoozeRule,
+  unsnoozeRule,
+}: ComponentOpts) => {
+  const [isEnabled, setIsEnabled] = useState<boolean>(item.enabled);
+  const [isSnoozed, setIsSnoozed] = useState<boolean>(isItemSnoozed(item));
+  useEffect(() => {
```

Comment:
This is necessary because we store `isEnabled` and `isSnooze` in state to prevent the loading spinner from flickering before switching a status. Consequently we have to manually listen for any outside updates of the `item`.

Here's a recording of what it looks like if we don't use these `useState`/`useEffect` combos and just relied on an `item` prop update to re-render the dropdown:
![Flicker](https://user-images.githubusercontent.com/1445834/159785062-400e0613-a816-43e8-85f6-97f20b10be54.gif)

Reply:
Hmm so it sounds like we're doing this to make the UI for responsive - in that we update the UI right after the user interaction while simultaneously sending the request to the server? That's the goal here? And without the `useState/useEffect`, we have to wait for the server to respond with the updated SO and if that takes a small amount of time (`500ms` to `1s`) we see this weird flickering of the loading icon, then the updated value?

Reply:
No the server has already responded with the updated SO, the wait time is just from React. Keeping things in `useState/useEffect` gives us control over when things are rendered.

Note that this same code already existed in the previous switch control.

Reply:
Okay we can defer to discuss this in the future as it's not a new patttern!